DllImportGenerator/DllImportGenerator/Analyzers/ConvertToGeneratedDllImportFixer.cs

```diff
@@ -0,0 +1,224 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+
+using static Microsoft.Interop.Analyzers.AnalyzerDiagnostics;
+
+namespace Microsoft.Interop.Analyzers
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public sealed class ConvertToGeneratedDllImportFixer : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(Ids.ConvertToGeneratedDllImport);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public const string NoPreprocessorDefinesKey = "ConvertToGeneratedDllImport";
+        public const string WithPreprocessorDefinesKey = "ConvertToGeneratedDllImportPreprocessor";
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            // Get the syntax root and semantic model
+            Document doc = context.Document;
+            SyntaxNode? root = await doc.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SemanticModel? model = await doc.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            if (root == null || model == null)
+                return;
+
+            // Nothing to do if the GeneratedDllImportAttribute is not in the compilation
```

Comment:
Wouldn't this require the type to already be in use? If so, how is this fixer bootstrapped? This would seem to make the concern about injecting the type rather than having it a part of .NETCoreApp even more relevant.

See https://github.com/dotnet/runtime/issues/46822#issuecomment-758986611

Reply:
It would need to be available, but not necessarily in use - e.g. could be in a referenced assembly, but not actually being used.

The fixer has to be installed via VSIX. It will have a chance to provide a fix when diagnostics corresponding to `FixableDiagnosticIds` are fired.

Your question does bring up the problem that source generators are not VSIX-provided, but analyzers and fixers are. Our analyzers do the same check for the attribute being in the compilation. This would mean that if we put the attribute in NETCoreApp, it would always be there (for NETCoreApp versions after it is added) - and our analyzer (installed through VSIX) would flag things even if the source generator isn't being used. This is mitigated by the fact that rule for migrating to `GeneratedDllImport` is currently not enabled by default, so the user has to explicitly enable it in their editorconfig.

Injecting the type in the source generator itself would help with that, since the type would only be available when the source generator is used.

Reply:
> Your question does bring up the problem that source generators are not VSIX-provided, but analyzers and fixers are.

This is a good point. Let's add a section in the design document about the fixer/analyzer/source generator scenario. You have answered by question/concern.

@chsienki @stephentoub Is there any guidance on how the fixer/analyzer/source generator scenarios align? How are they installed as a unit? Is a fixer from the command line going to be a thing yet? If I recall it was mentioned in an offline chat I had with someone but I am not sure where that ended up.